### PR TITLE
JIT: Remove double-negation in during morph phase

### LINKED ISSUE #27442: JIT: Optimize -(-x) to x
JIT should be able to eliminate double negations, as far as I understand it's safe to do so for any signed integers and floats:

```csharp
static float M1(float a)
{
    return -(-a); // e.g. after inlining
}

static int I1(int a)
{
    return -(-a);
}

/*
\--*  NEG       float 
   \--*  NEG       float 
      \--*  LCL_VAR   float  V00 arg0
*/
```
Current codegen:
```asm
; Method M1(float):float
G_M37219_IG01:
       vzeroupper 
G_M37219_IG02:
       vmovss   xmm1, dword ptr [reloc @RWD00]
       vxorps   xmm0, xmm1
       vmovss   xmm1, dword ptr [reloc @RWD00]
       vxorps   xmm0, xmm1
G_M37219_IG03:
       ret      
RWD00  dd	80000000h
; Total bytes of code: 28


; Method I1(int):int
G_M37095_IG01:
G_M37095_IG02:
       mov      eax, ecx
       neg      eax
       neg      eax
G_M37095_IG03:
       ret      
; Total bytes of code: 7
```

Expected codegen:
```asm
; Method M1(float):float
G_M37219_IG01:
       vzeroupper 
G_M37219_IG02:
       ret      
; Total bytes of code: 4


; I1(int):int
G_M37095_IG01:
G_M37095_IG02:
       mov      eax, ecx
G_M37095_IG03:
       ret      
; Total bytes of code: 3
```
A double negation might confuse the optimization introduced in https://github.com/dotnet/coreclr/pull/27060

### PR DESCRIPTION
**Pull request by damageboy**
- fixes #27442
- Co-authored with @EgorBo holding my hand :)

### CODE REVIEW

**Comment by EgorBo** [comment]
I think you still need to check the overflow flag, the following case should throw an exception, shouldn't it?
```csharp
checked
{
    int a = int.MaxValue;
    Console.WriteLine(a == -(-a));

    int b = int.MinValue;
    Console.WriteLine(b == -(-b));
}
```

**Comment by mikedn** [comment]
> I think you still need to check the overflow flag

There's no overflow flag on `GT_NEG`.

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -12791,6 +12791,12 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
 
         case GT_NOT:
         case GT_NEG:
+            // Remove double negation
+            if (oper == GT_NEG && op1->OperIs(GT_NEG))
```
Comment (erozenfeld) [inline src/jit/morph.cpp]: Please change 
`if (oper == GT_NEG && op1->OperIs(GT_NEG))`
to
`if ((oper == GT_NEG) && op1->OperIs(GT_NEG))`
per our convention.

**Comment by erozenfeld** [comment]
@EgorBo The C# snippet in your comment will result in sub.ovf msil instructions, not neg. As @mikedn pointed out `neg` will never result in an overflow. From ECMA-335:
```
The neg instruction negates value and pushes the result on top of the stack. The return type is the
same as the operand type.
Negation of integral values is standard twos-complement negation. In particular, negating the
most negative number (which does not have a positive counterpart) yields the most negative
number. To detect this overflow use the sub.ovf instruction instead (i.e., subtract from 0). 
```

File: src/jit/morph.cpp (hunk 1 above)
Comment (damageboy) [reply src/jit/morph.cpp]: Fixed!

**Review by sandreenko — COMMENTED** [review]
The change looks good with 1 nit that is optional, thank you, @damageboy.

Could you please collect jit asm diffs so we now how many methods are affected and make sure there are no unexpected regressions?

In order to do that refer to https://github.com/dotnet/jitutils/blob/master/doc/diffs.md, the final command will look like:
```
C:\git\tools\jitutils\bin\jit-diff.bat diff -b "F:\git\coreclr\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root_base" -d "F:\git\coreclr\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root_diff" -o F:\diffs --core_root F:\git\coreclr\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root -c
```

File: src/jit/morph.cpp (hunk 1 above)
Comment (sandreenko) [inline src/jit/morph.cpp]: Nit: It is strange to see an `if` on `switch` operand.
I would do 
```
case GT_NEG:
	if (op1->OperIs(GT_NEG))
	{
	    // Remove double negation
		GenTree* child = op1->AsOp()->gtGetOp1();
		return child;
	}
	__fallthrough;
case GT_NOT:
	old code.

```

Comment (damageboy) [reply src/jit/morph.cpp]: Sure thing, it wasn't clear to me what the preferred style is, to keep the commonality or separate cases.

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.

**Comment by BruceForstall** [comment]
@damageboy Are you going to open a PR for this in the https://github.com/dotnet/runtime repo? (Or did you already do so and I missed it?)

**Comment by damageboy** [comment]
Yes! I am planning to do this sometime over this weekend :)

**Pull request closed without merging**